### ISSUE TITLE
Internal compiler error

### ISSUE THREAD

**Issue by yuriv** [eggs-cpp/variant#31]
I'll try to compile eggs::variant with gcc 7.3.1 (devtoolset-7) on linux and got folowing error:

```
In file included from <...>/variant/include/eggs/variant.hpp:13:0,
                          from <...>/variant/test/swap.cpp:8:
<...>/variant/include/eggs/variant/variant.hpp: In substitution of ‘template<class U, class NoCopy, class NoTag, long unsigned int I, class T, typename std::enable_if<(std::is_constructible<T, U>::value && (! std::is_convertible<U, T>::value)), bool>::type <anonymous> > constexpr eggs::variants::variant<int, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >::variant(U&&) [with U = int; NoCopy = void; NoTag = void; long unsigned int I = <missing>; T = <missing>; typename std::enable_if<(std::is_constructible<T, U>::value && (! std::is_convertible<U, T>::value)), bool>::type <anonymous> = <missing>]’:
<...>/variant/test/swap.cpp:116:46:   required from here
<...>/variant/include/eggs/variant/variant.hpp:696:13: internal compiler error: unexpected expression ‘I’ of kind template_parm_index
           , typename T = typename detail::at_index<
             ^~~~~~~~

Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
Preprocessed source stored into /tmp/ccYpdTC0.out file, please attach this to your bugreport.**
```

With gcc 6 it compiles and works fine.

machine:
     centos-release-7-5.1804.el7.centos.2.x86_64
     kernel: Linux 3.10.0-862.9.1.el7.x86_64

gcc --version:
    gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)

**Comment by K-ballo** [comment]
An internal compiler error is a compiler bug, always, you should report it. Meanwhile, we can try to find a workaround for it. Is `swap` the only test hitting this issue?

The issue does not reproduce with gcc 7.3.0, all tests compile and work fine. I'll see to get 7.3.1 to try again.

**Comment by K-ballo** [comment]
I was able to reproduce with 7.3.1 on CentOS 7. Looking for a workaround now.

**Comment by K-ballo** [comment]
The following patch is sufficient to get all tests to compile:

```diff

**Comment by K-ballo** [comment]
The following patch is sufficient to workaround the issue and make all tests compile:

```diff
include/eggs/variant/variant.hpp | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/include/eggs/variant/variant.hpp b/include/eggs/variant/variant.hpp
index c4f48c0..f7547f0 100644
--- a/include/eggs/variant/variant.hpp
+++ b/include/eggs/variant/variant.hpp
@@ -650,7 +650,9 @@ namespace eggs { namespace variants
           , std::size_t I = detail::index_of_best_match<
                 U&&, detail::pack<Ts...>>::value
           , typename T = typename detail::at_index<
-                I, detail::pack<Ts...>>::type
+                detail::index_of_best_match<
+                    U&&, detail::pack<Ts...>>::value,
+                detail::pack<Ts...>>::type
           , typename std::enable_if<
                 std::is_constructible<T, U>::value
              && std::is_convertible<U, T>::value
@@ -694,7 +696,9 @@ namespace eggs { namespace variants
           , std::size_t I = detail::index_of_explicit_match<
                 U&&, detail::pack<Ts...>>::value
           , typename T = typename detail::at_index<
-                I, detail::pack<Ts...>>::type
+               detail::index_of_explicit_match<
+                    U&&, detail::pack<Ts...>>::value,
+                detail::pack<Ts...>>::type
           , typename std::enable_if<
                 std::is_constructible<T, U>::value
              && !std::is_convertible<U, T>::value
@@ -946,7 +950,9 @@ namespace eggs { namespace variants
           , std::size_t I = detail::index_of_best_match<
                 U&&, detail::pack<Ts...>>::value
           , typename T = typename detail::at_index<
-                I, detail::pack<Ts...>>::type
+                detail::index_of_best_match<
+                    U&&, detail::pack<Ts...>>::value,
+                detail::pack<Ts...>>::type
           , typename std::enable_if<
                 std::is_assignable<T&, U>::value
              && std::is_constructible<T, U>::value, bool>::type = true
@@ -974,7 +980,9 @@ namespace eggs { namespace variants
           , std::size_t I = detail::index_of_best_match<
                 U&&, detail::pack<Ts...>>::value
           , typename T = typename detail::at_index<
-                I, detail::pack<Ts...>>::type
+                detail::index_of_best_match<
+                    U&&, detail::pack<Ts...>>::value,
+                detail::pack<Ts...>>::type
           , typename std::enable_if<
                 !std::is_assignable<T&, U>::value, bool>::type = false
         >
```

The workaround is unsatisfactory. I'll try to get to the root of the compiler bug.

**Comment by K-ballo** [comment]
Considering 7.3.1 is a development snapshot, and that the code works with both 7.3.0 and 8.x, I am not looking at making source changes for this. If it does not get fixed in time for 7.4.0 I'll reconsider.

Please report the regression to GCC and share the link here so we may track it's progress.

**Comment by K-ballo** [comment]
> Please report the regression to GCC and share the link here so we may track it's progress.

Looks like this is https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84785, and the fix has already landed in the gcc-7-branch.
